### ISSUE TITLE
Double-slash in pathnames if their path.generator is false'd 

### ISSUE THREAD

**Issue by IllyaMoskvin** [nWidart/laravel-modules#368]
**tl;dr** setting certain parameters in `path.generator` to `false` results in double forward slashes being inserted throughout the generated code, due to empty variable in concatenation.

If I set e.g. `path.generator.provider` to `false` (or `null` or an empty string), with the intention that the default `ModelServiceProvider` should live in the root directory of the module, I get the following readout in the console:

```bash
php artisan module:make Blog

[snip irrelevant stuff]

Created : /[...]/repo/Modules/Blog//BlogServiceProvider.php
```

Note the double `//`. Now, my system recovered, and the file was created in the root of the module as expected. The same thing happened to `BlogDatabaseSeeder` and `BlogController` upon further testing, if I false'd out `seeder` and `controller` respectively.

Likewise, when I set `path.generator.config` to `false`, my `BlogServiceProvider::registerConfig()` looked like this:

```
    protected function registerConfig()
    {
        $this->publishes([
            __DIR__.'/..//config.php' => config_path('blog.php'),
        ], 'config');
        $this->mergeConfigFrom(
            __DIR__.'/..//config.php', 'blog'
        );
    }
```

I don't know if the extra `/` will affect things, since I fixed it manually before running.

If I false'd out `path.generator.view`, I had a line in `registerViews()` like so:

```
        $sourcePath = __DIR__.'/../';
```

Same thing for `path.generator.lang` and `registerTranslations()`:

```
            $this->loadTranslationsFrom(__DIR__ .'/../', 'blog');
```

There might be other examples I missed, but that all I've found thus far.

Now, I haven't tested it before fixing these mistakes, but something seems off about this.

For example, will `loadTranslationsFrom` fail if there's no translations to be found in the module root? (Technically, if both `provider` and `lang` are false'd out as above, it'll look in the directory above the module root.) Likewise, will `mergeConfigFrom` fail due to the extra forward slash?

I don't disagree that we should emphasize to users the need to edit their generated `ModuleServiceProvider`'s to get things working properly within the context of their application.

However, I think it might be nice to infer what methods should be added to those providers based off the app's `path.generator` settings. For example, if I set `'lang' => false`, I would expect the `registerTranslations` block to be removed from my generated `Provider`. Likewise for views, factories, and config.

I know this won't solve some of the issues mentioned above (e.g. the combination of setting both `provider` and `lang` to false), and there's still the need to differentiate between `false` and `''` in some cases, but it might be a part of the solution.

**Comment by nWidart** [comment]
This is normal yes.
If it's set to false, the path won't be correct. If you do want the path to be correct it needs to be a correct path and not `false`.

**Comment by IllyaMoskvin** [comment]
Would you be interested in a PR to fix the issue?

What are your thoughts re: omitting method blocks from the generated `Provider` if the corresponding `path.generator` setting is `false`? If you prefer, I can open a separate issue.

**Comment by nWidart** [comment]
Hm, I'm not sure this needs fixing as it's pretty obvious if you set it to false.

Though it could be nice to disable a bunch of folders not needed during module generation, but when generating a file later that the missing folder is generated if needed.
The only solution I see for this is having that whole config option duplicated.

**Comment by nWidart** [comment]
Maybe having something like this at the root of the config file, and comment/remove parts that shouldn't be generated on initial `module:make` 🤔 

```php
'generate' => [
    'assets',
    'config',
    'commands',
    'events',
    'listeners',
    'migrations',
    'factories',
    'models',
    'repositories',
    'seeders',
    'controllers',
    'middleware',
    'requests',
    'providers',
    'lang',
    'views',
    'policies',
    'rules',
    'tests',
    'jobs',
    'emails',
    'notifications',
    'resources',
],
```
Though it's pretty big.

**Comment by IllyaMoskvin** [comment]
I have to disagree that it's obvious for new users. When I first found the project, I read the [Introduction](https://nwidart.com/laravel-modules/v2/introduction) and thought, "Man, that's a lot of folders to delete, and I don't know what is and is not required." Then I looked at the [config](https://github.com/nWidart/laravel-modules/blob/master/config/config.php) file and thought, "Oh cool, there's a dedicated setting to customize which folders get generated."

It took me a while to realize that not _everything_ can be controlled via this setting, e.g. if I don't want a config, I have to set `paths.generator.config` to false, comment out the `scaffold/config` stub file, and delete `registerConfig` from the generated Provider. This was surprising, and felt somewhat buggy. Seeing empty or malformed paths where one might expect to see nothing was just a part of that initial impression.

I don't think it's a big deal for users who have worked with the library for a while. After all, it's not like most folks will be generating new modules every day. But it makes migrating an existing app to a modular architecture a bit painful, and it's surprising to newcomers.

Sorry this took a while to write. I like the approach above! I'm not sure how it would interact with the file stubs, e.g. what might happen if someone comments out `config` from the list, but keeps the `scaffold/config` stub?

Also, if the above approach was implemented, would that mean that the semantics for setting a generator path to `false` would change to something more like "Set the value to false to place those resources in the module's root folder."..? That might clear up some of the confusion.

**Comment by nWidart** [comment]
Yeah, I see what you mean.

I was discussing this with @julientant , and instead of duplicating the keys, it could be something like this:

```php
'assets' => ['path' => 'Assets' , 'generate' => false],
```

In regards to files that's something else entirely I think. If the config folder isn't generated, the config file won't be in the right location. Though I think that's to be expected.

**Comment by IllyaMoskvin** [comment]
Seems like an elegant solution. Are there worries about backward compatibility?

I think you're right on the files. Given that people can change the stubs entirely, it'd be a bad idea to decide whether or not the `registerConfig` block is added to the default `Provider` based on the presence of `scaffold/config`.

**Comment by nWidart** [comment]
As is this would be a breaking change yeah. Though I'm thinking to abstract this to a class that could check if array or not to keep backwards compatibility.

**Comment by JulienTant** [comment]
That's the way to go ! ;)

**Comment by IllyaMoskvin** [comment]
Right on, sounds great! Let me know if you guys want a PR targeting any of the things that have been discussed. I should have some time next week to dig deeper.

Related question: if `controller` is set to `'generate' => false`, should the default `ModuleController.php` still be generated?

Based on [ModuleGenerator.php#L261](https://github.com/nWidart/laravel-modules/blob/70a113e6302e196ffff039ba71795a127cb6ef26/src/Generators/ModuleGenerator.php#L261) and [L322](https://github.com/nWidart/laravel-modules/blob/70a113e6302e196ffff039ba71795a127cb6ef26/src/Generators/ModuleGenerator.php#L322), it's an all-or-nothing affair right now, i.e. there's no way to disable controller generation without also disabling the seeder and provider. This `generate` param might be the right indicator to check before generating a controller with `module:make`.

Likewise for `seeder`. I imagine the `ModuleServiceProvider` is required and should be always generated to avoid confusion, i.e. I can see folks thinking, "I don't need custom Providers," and then wondering what happened to their `ModuleServiceProvider`.

**Comment by nWidart** [comment]
Yea those are good ideas, but too breaking for now.

This test is green currently https://gist.github.com/nWidart/a3d1dc77279c2ad66dbb6ad99e33b009

**Comment by nWidart** [comment]
Change has been applied 😱  https://github.com/nWidart/laravel-modules/commit/2eb38ceca2e8735032f2d1e621d66c3ff3727280

New defaults has been set as well, think this makes more sense: https://github.com/nWidart/laravel-modules/commit/ccda05a712a8d1e27471a126f11aaa287a8e90c8

**Comment by nWidart** [comment]
Tagged on 2.5

**Comment by IllyaMoskvin** [comment]
Awesome, thank you! I'm really impressed with how fast these issues have been addressed.

Later today, I'll go ahead and spin out a couple things discussed above into their own issues, so they could be discussed further and (maybe) tagged for a future milestone.
